### PR TITLE
Add netcdf-c 4.9.0 and netcdf-fortran 4.6.0

### LINKED ISSUE #295: netcdf-c and netcdf-fortran can be updated, with special measures to enable zstandard compression
netcdf-c-4.9.0 and netcdf-fortran-4.6.0 are out! ;-)

In order for zstandard compression to work, the zstandard library must be installed before netcdf-c is built.

Also the configure option --with-plugin-dir=/some/dir.

Then the environment variable HDF5_PLUGIN_PATH must be set to /some/dir, for netcdf-fortran and every application that wants to use zstandard.

That is, if some rando netCDF application calls nc_def_var_zstandard() (or its fortran wrappers), the must have the HDF5_PLUGIN_PATH environment variable set correctly. or when they run that code they will get a "filter not found" error.

So this may present a challenge for spack. @kgerheiser you are now much more familar with spack than I am - what do you think is the best way to do this?

### PR DESCRIPTION
**Pull request by kgerheiser**
With v4.9.0 netcdf-c introduces zstandard compression option which is added as a variant.

@edwardhartnett I've got NetCDF detecting zstd, now where is `HDF5_PLUGIN_PATH` supposed to point? I thought that would be a plugin directory in the NetCDF installation.

- waiting on https://github.com/spack/spack/pull/31953
- fixes https://github.com/NOAA-EMC/spack-stack/issues/295

### CODE REVIEW

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 1]
```diff
@@ -59,7 +60,7 @@ class NetcdfC(AutotoolsPackage):
     patch('4.7.3-spectrum-mpi-pnetcdf-detect.patch', when='@4.7.3:4.7.4 +parallel-netcdf')
 
     # See https://github.com/Unidata/netcdf-c/pull/2293
-    patch('4.8.1-no-strict-aliasing-config.patch', when='@4.8.1:')
+    patch('4.8.1-no-strict-aliasing-config.patch', when='@4.8.1')
```
Comment (kgerheiser) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]: This was fixed in 4.9.0

**Comment by climbfuji** [comment]
This looks good, what about the `HDF5_PLUGIN_PATH` question?

**Comment by kgerheiser** [comment]
> What about the HDF5_PLUGIN_PATH question?

Need @edwardhartnett input on that

**Comment by edwardhartnett** [comment]
HDF5_PLUGIN_PATH must point to the same place that was given as an argument to --with-plugin-dir= option of netCDF configure.

It can a directory called plugins in the install lib directory, for example. But whatever --with-plugin-dir= is set to, that's what HDF5_PLUIN_PATH must include. (Actually it can be a semi-colon separated list of paths...)

So HDF5_PLUGIN_PATH is where HDF5 applications (like the UFS) will look for the zstandard plugin at RUNTIME (not compile time).

**Comment by kgerheiser** [comment]
Oh, right. I didn't pass that option to configure. I would set that to `self.prefix.plugins` and install it along the rest of NetCDF.

**Comment by kgerheiser** [comment]
Adding `--with-plugin-dir=` and I see it populated with the `zstd` filters.

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 2]
```diff
@@ -239,6 +248,10 @@ def configure_args(self):
 
         return config_args
 
+    def setup_run_environment(self, env):
+        if self.spec.satisfies('+zstd'):
+            env.set('HDF5_PLUGIN_PATH', self.prefix.plugins)
```
Comment (kgerheiser) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]: Should append to HDF5_PLUGIN_PATH instead of setting it.

**Review by climbfuji — COMMENTED** [review]
@kgerheiser Is this PR ready?

**Comment by climbfuji** [comment]
I guess we wait to see the outcome of your spack PR (50 github comments and counting), and then merge the "final" product into our fork ...

**Comment by kgerheiser** [comment]
I would say it's ready now. I addressed all the issues in the original PR.

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 3]
```diff
@@ -80,172 +88,192 @@ class NetcdfC(AutotoolsPackage):
     #         description='Enable CDM Remote support')
 
     # The patch for 4.7.0 touches configure.ac. See force_autoreconf below.
-    depends_on('autoconf', type='build', when='@4.7.0,main')
-    depends_on('automake', type='build', when='@4.7.0,main')
-    depends_on('libtool', type='build', when='@4.7.0,main')
+    depends_on("autoconf", type="build", when="@4.7.0,main")
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:91]: This type of changes (coming from the major cleanup that Todd did in the authoritative spack repo) will make it difficult for us to pull in selective updates until we get the cleanup commit itself merged across.

**Review by climbfuji — COMMENTED** [review]
I noticed the style tests are failing. Otherwise it's looking good,  although it would have been nice to get Todds huge format PR merged in first so that 90% of the changes in this PR wouldn't show up. But it's fine the way it is.

File: var/spack/repos/builtin/packages/netcdf-c/package.py (hunk 3 above)
Comment (kgerheiser) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:91]: Yeah, I didn't notice it at first, and didn't know that this was in v0.18.1. I'll rebase my PR on top of that.

**Review by climbfuji — APPROVED** [review]
Nice work. Maybe a little more cleanup, as discussed, but good to go.

**Comment by edwardhartnett** [comment]
Great work all! @kgerheiser thanks for the assist and good luck in your new role. I'm sure you'll have those people at Intel sorted out soon...

image

**Comment by climbfuji** [comment]
We should have grabbed his branch before it got deleted, now we have to do all of the work again ... same for the cylc packages.

**Comment by edwardhartnett** [comment]
@kgerheiser can you undelete this?

**Comment by kgerheiser** [comment]
Yes, I re-created the repo and branches.

**Comment by climbfuji** [comment]
Oh great, thank you. I'll clone your branches and put them in the spack/spack-stack repos so that we have them.

**Pull request closed without merging**